Tools/Scripts/libraries/webkitcorepy/webkitcorepy/file_lock.py

```diff
@@ -22,60 +22,86 @@
 # SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 import os
+import re
 import sys
 import time
 
+from webkitcorepy import string_utils
+
 if sys.platform.startswith('win'):
     import msvcrt
-else:
-    import fcntl
 
 
 class FileLock(object):
+    INTEGER_RE = re.compile(r'^\d+$')
+    USE_WINDOWS = sys.platform.startswith('win')
+    USE_EXLOCK = not USE_WINDOWS and getattr(os, 'O_EXLOCK', False)
+
+    @classmethod
+    def is_process_running(cls, pid):
+        try:
+            os.kill(pid, 0)
+            return True
+        except OSError:
+            return False
 
     def __init__(self, path, timeout=20):
         self.path = path
         self.timeout = timeout
         self._descriptor = None
+        self._depth = 0
 
     @property
     def acquired(self):
         return bool(self._descriptor)
 
     def acquire(self):
         if self._descriptor:
+            self._depth += 1
             return True
 
-        descriptor = os.open(self.path, os.O_TRUNC | os.O_CREAT)
+        if not self.USE_EXLOCK and not self.USE_WINDOWS and os.path.exists(self.path):
+            with open(self.path) as file:
+                pid = file.readline().strip()
+            if self.INTEGER_RE.match(pid) and not self.is_process_running(int(pid)):
+                os.unlink(self.path)
```

Comment:
There is a race here. If two processes get to this line, one can then get all the way to line 79 and create a new lock file, and then another process will delete it. I don't know if/how this is normally solved.

Reply:
That's true, although I'm comfortable leaving this race condition in place, given I'm not familiar with the tools other systems we might be running on have access to.

It's worth pointing out that this race condition requires 2 things. The first is that a previous process was actually killed, because if a process is merely interrupted, we clean everything up as we unwind the stack. The second is that we have two processes then racing the initial lock at the same time. It seems like our other option would be to avoid the race condition by refusing to acquire the lock because we can't check if the process which originally acquired it is live. That seems strictly worse, because we still fail in your scenario, but now we also fail if we only have one process running after a previous process has been killed.